sdk/rapid/rapid.py

```diff
@@ -413,3 +413,130 @@ def update_subject_permissions(self, subject_id: str, permissions: list[str]):
+        response = requests.post(
+            url,
+            headers=self.generate_headers(),
+            data=json.dumps(
+                {"username": user_name, "email": user_email, "permissions": user_permissions}
+            ),
+            timeout=TIMEOUT_PERIOD,
+        )
+        data = json.loads(response.content.decode("utf-8"))
+        if response.status_code == 201:
+            return data
+        elif response.status_code == 400:
+            raise SubjectAlreadyExistsException(
+                f"The user {user_name} already exists"
+            )
+        raise InvalidPermissionsException(
+            "One or more of the provided permissions is invalid or duplicated"
+        )
+    
+    def delete_user(self, user_name: str, user_id: str):
+        """
+        Deletes a client from the API based on their id
+
+        Args:
+            client_id (str): The id of the client to delete.
+
+        Raises:
+            rapid.exceptions.SubjectNotFoundException: If the client does not exist.
+        """
+        url = f"{self.auth.url}/user"
+        response = requests.delete(
+            url,
+            headers=self.generate_headers(),
+            data=json.dumps(
+                {"username": user_name, "user_id": user_id}
+            ),
+            timeout=TIMEOUT_PERIOD,
+        )
+        data = json.loads(response.content.decode("utf-8"))
+        if response.status_code == 200:
+            return data
+
+        raise SubjectNotFoundException(
+            f"Failed to delete user with id: {user_id}, ensure it exists."
+        )
+    
+    def list_subjects(self):
+        """
+        List all current subjects within rAPId instance.
+
+        Returns:
+            A JSON response of the API's response.
+        """
+        response = requests.get(
+            f"{self.auth.url}/subjects",
+            headers=self.generate_headers(),
+            timeout=TIMEOUT_PERIOD,
+        )
+        return json.loads(response.content.decode("utf-8"))
+    
+    def list_layers(self):
+        """
+        List all current layers within rAPId instance.
+
+        Returns:
+            A JSON response of the API's response.
+        """
+        response = requests.get(
+            f"{self.auth.url}/layers",
+            headers=self.generate_headers(),
+            timeout=TIMEOUT_PERIOD,
+        )
+        return json.loads(response.content.decode("utf-8"))
+    
+    def list_protected_domains(self):
+        """
+        List all current protected domains within rAPId instance.
+
+        Returns:
+            A JSON response of the API's response.
+        """
+        response = requests.get(
+            f"{self.auth.url}/protected_domains",
+            headers=self.generate_headers(),
+            timeout=TIMEOUT_PERIOD,
+        )
+        return json.loads(response.content.decode("utf-8"))
+    
+    def delete_dataset(self, layer: str, domain: str, dataset: str):
```

Comment:
How does this function handle incorrect permissions for deleting a dataset?

Reply:
I will cover it.
However, there is an issue with API, it returns 202 when dataset doesn't exists but layer is correct. It only returns 400 when layer is incorrect.

Reply:
Fixed.